File: Februrary 2025/24.py
class Solution:
    def calculateSpan(self, arr):
        # Initialize a stack and an array to store the spans
        stack = []
        span = [0] * len(arr)

        # Iterate through the list of prices
        for i in range(len(arr)):
            # While stack is not empty and the current price is greater than
            # the price at the index stored at the top of the stack
            while stack and arr[i] >= arr[stack[-1]]:
                stack.pop()
            # If stack is empty, it means the current price is greater than all previous prices
            if not stack:
                span[i] = i + 1
            else:
                # Calculate the span by subtracting the index of the last higher price from the current index
                span[i] = i - stack[-1]
            # Push the current index to the stack
            stack.append(i)

        return span

File: Februrary 2025/24.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
  public:
    vector<int> calculateSpan(vector<int>& arr) {
        // Initialize a stack and a vector to store the spans
        stack<int> s;
        vector<int> span(arr.size(), 0);

        // Iterate through the list of prices
        for (int i = 0; i < arr.size(); i++) {
            // While stack is not empty and the current price is greater than
            // the price at the index stored at the top of the stack
            while (!s.empty() && arr[i] >= arr[s.top()]) {
                s.pop();
            }
            // If stack is empty, it means the current price is greater than all previous prices
            if (s.empty()) {
                span[i] = i + 1;
            } else {
                // Calculate the span by subtracting the index of the last higher price from the current index
                span[i] = i - s.top();
            }
            // Push the current index to the stack
            s.push(i);
        }

        return span;
    }
};
